syncer/streamer_controller.go

```diff
@@ -0,0 +1,412 @@
+
+	fromDB *UpStreamConn
+
+	uuidSuffix string
+
+	closed bool
+}
+
+// NewStreamerController creates a new streamer controller
+func NewStreamerController(tctx *tcontext.Context, syncCfg replication.BinlogSyncerConfig, fromDB *UpStreamConn, binlogType BinlogType, localBinlogDir string, timezone *time.Location, beginPos mysql.Position) (*StreamerController, error) {
+	streamerController := &StreamerController{
+		initBinlogType:    binlogType,
+		currentBinlogType: binlogType,
+		syncCfg:           syncCfg,
+		localBinlogDir:    localBinlogDir,
+		timezone:          timezone,
+		fromDB:            fromDB,
+		closed:            false,
+	}
+
+	err := streamerController.UpdateServerIDAndResetReplication(tctx, beginPos)
+	if err != nil {
+		streamerController.Close(tctx)
+		return nil, err
+	}
+
+	return streamerController, nil
+}
+
+// ResetReplicationSyncer reset the replication
+func (c *StreamerController) ResetReplicationSyncer(tctx *tcontext.Context, pos mysql.Position) (err error) {
+	c.Lock()
+	defer c.Unlock()
+
+	return c.resetReplicationSyncer(tctx, pos)
+}
+
+func (c *StreamerController) resetReplicationSyncer(tctx *tcontext.Context, pos mysql.Position) (err error) {
+	uuidSameWithUpstream := true
+
+	// close old streamerProducer
+	if c.streamerProducer != nil {
+		switch t := c.streamerProducer.(type) {
+		case *remoteBinlogReader:
+			c.closeBinlogSyncer(tctx, t.reader)
+		case *localBinlogReader:
+			// check the uuid before close
+			uuidSameWithUpstream, err = c.checkUUIDSameWithUpstream(pos, t.reader.GetUUIDs())
+			if err != nil {
+				return err
+			}
+			t.reader.Close()
+		default:
+			// some other producers such as mockStreamerProducer, should not re-create
+			c.streamer, err = c.streamerProducer.generateStreamer(pos)
+			return err
+		}
+	}
+
+	if c.initBinlogType == LocalBinlog && c.meetError {
+		// meetError is true means meets error when get binlog event, in this case use remote binlog as default
+		if !uuidSameWithUpstream {
+			// if the binlog position's uuid is different from the upstream, can not switch to remote binlog
+			tctx.L().Warn("may switch master in upstream, so can not switch local to remote")
+		} else {
+			c.currentBinlogType = RemoteBinlog
+			tctx.L().Warn("meet error when read from local binlog, will switch to remote binlog")
+		}
+	}
+
+	if c.currentBinlogType == RemoteBinlog {
+		c.streamerProducer = &remoteBinlogReader{replication.NewBinlogSyncer(c.syncCfg), tctx, false}
+	} else {
+		c.streamerProducer = &localBinlogReader{streamer.NewBinlogReader(tctx, &streamer.BinlogReaderConfig{RelayDir: c.localBinlogDir, Timezone: c.timezone})}
+	}
+
+	c.streamer, err = c.streamerProducer.generateStreamer(pos)
+	return err
+}
+
+// RedirectStreamer redirects the streamer's begin position or gtid
+func (c *StreamerController) RedirectStreamer(tctx *tcontext.Context, pos mysql.Position) error {
+	c.Lock()
+	defer c.Unlock()
+
+	tctx.L().Info("reset global streamer", zap.Stringer("position", pos))
```

Comment:
How about removing `global` here? or even we can use `ResetReplicationSyncer` instead `RedirectStreamer`?

Reply:
I update the log in f05ac87